base/compiler/tfuncs.jl

```diff
@@ -1407,39 +1437,13 @@ function apply_type_nothrow(@specialize(lattice::AbstractLattice), argtypes::Arr
     u = headtype
     for i = 2:length(argtypes)
         isa(u, UnionAll) || return false
-        ai = widenconditional(argtypes[i])
-        if ⊑(lattice, ai, TypeVar) || ai === DataType
-            # We don't know anything about the bounds of this typevar, but as
-            # long as the UnionAll is not constrained, that's ok.
-            if !(u.var.lb === Union{} && u.var.ub === Any)
-                return false
-            end
-        elseif (isa(ai, Const) && isa(ai.val, Type)) || isconstType(ai)
-            ai = isa(ai, Const) ? ai.val : (ai::DataType).parameters[1]
-            if has_free_typevars(u.var.lb) || has_free_typevars(u.var.ub)
-                return false
-            end
-            if !(u.var.lb <: ai <: u.var.ub)
```

Comment:
Doesn't the deletion of this case break lots of easy cases like `Complex{Int}`?

Reply:
Easy cases never get here, since they are Const above and hit the early out. This only runs if something is unknown about the parameters